[css-anchor-position-1] Implement `anchor-center` alignment value

Scony opened the pull request:
#### a84350ee19f346b0c7515b2c84464b055ac4c954
<pre>
[css-anchor-position-1] Implement `anchor-center` alignment value
<a href="https://bugs.webkit.org/show_bug.cgi?id=275451">https://bugs.webkit.org/show_bug.cgi?id=275451</a>

Reviewed by Antti Koivisto.

This change implements `anchor-center` CSS value as specified in:
<a href="https://drafts.csswg.org/css-anchor-position-1/#anchor-center">https://drafts.csswg.org/css-anchor-position-1/#anchor-center</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-htb-htb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-htb-vrl-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-vrl-htb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-vrl-vrl-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computePositionedLogicalWidth const):
(WebCore::RenderBox::computePositionedLogicalHeight const):
(WebCore::isObjectAncestorContainerOf):
(WebCore::findClosestCommonContainer):
(WebCore::RenderBox::computeAnchorCenteredPosition const):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::defaultAnchor const):
(WebCore::RenderElement::defaultAnchorRenderer const):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::alignmentOffset):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::columnAxisPositionForGridItem const):
(WebCore::RenderGrid::rowAxisPositionForGridItem const):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::computeAnchorRectRelativeToContainingBlock):
(WebCore::Style::computeInsetValue):
(WebCore::Style::computeAnchorRectRelativeToContainingBlock): Deleted.
* Source/WebCore/style/AnchorPositionEvaluator.h:

Canonical link: <a href="https://commits.webkit.org/287144@main">https://commits.webkit.org/287144@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/01ac09f52e622085ef97fe65033fa856ebe3d054

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78460 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57505 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31842 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83121 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29725 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80593 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66656 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5786 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61447 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19365 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81527 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51460 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/69285 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41760 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48806 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25299 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28062 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69903 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25583 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84487 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5825 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3968 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69718 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5986 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67451 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68928 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17188 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12950 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11328 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5772 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/9118 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5761 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9194 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7547 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1607,6 +1608,31 @@ bool RenderElement::isVisibleInViewport() const
+{
+    if (!element())
+        return nullptr;
+    Style::AnchorPositionedStates& anchorPositionedStates = document().styleScope().anchorPositionedStates();
```
```suggestion
    auto& anchorPositionedStates = document().styleScope().anchorPositionedStates();
```

nt1m commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1607,6 +1608,31 @@ bool RenderElement::isVisibleInViewport() const
+    auto anchoringStateLookupResult = anchorPositionedStates.find(*element());
+    if (anchoringStateLookupResult == anchorPositionedStates.end() || !anchoringStateLookupResult->value)
+        return nullptr;
+    const Style::AnchorPositionedState& anchoringState = *anchoringStateLookupResult->value;
```
```suggestion
    const auto& anchoringState = *anchoringStateLookupResult->value;
```

nt1m commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1607,6 +1608,31 @@ bool RenderElement::isVisibleInViewport() const
+RenderElement* RenderElement::defaultAnchorRenderer() const
+{
+    if (auto* aDefaultAnchor = defaultAnchor())
+        return aDefaultAnchor->renderer();
```
```suggestion
    if (auto* defaultAnchor = this->defaultAnchor())
        return defaultAnchor->renderer();
```

nt1m commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5906,4 +5943,81 @@ void RenderBox::removeShapeOutsideInfo()
+
+static const RenderBlock* findClosestCommonContainer(const RenderElement& elementA, const RenderElement& elementB)
+{
+    const RenderBlock* closestCommonContainer = dynamicDowncast<RenderBlock>(&elementA);
```
```suggestion
    CheckedPtr closestCommonContainer = dynamicDowncast<RenderBlock>(&elementA);
```

nt1m commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -4005,6 +4006,21 @@ void RenderBox::computePositionedLogicalWidth(LogicalExtentComputedValues& compu
     Length logicalRightLength = style().logicalRight();
 
+    // https://drafts.csswg.org/css-anchor-position-1/#anchor-center
+    RenderBoxModelObject* defaultAnchorBox = nullptr;
```
```suggestion
    RenderBoxModelObject* defaultAnchorBox;
```

Scony replied:
I've reworked that part a bit

nt1m commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -4418,6 +4437,21 @@ void RenderBox::computePositionedLogicalHeight(LogicalExtentComputedValues& comp
     Length logicalBottomLength = styleToUse.logicalBottom();
 
+    // https://drafts.csswg.org/css-anchor-position-1/#anchor-center
+    RenderBoxModelObject* defaultAnchorBox = nullptr;
```
```suggestion
    RenderBoxModelObject* defaultAnchorBox;
```

Scony replied:
I've reworked that part a bit

nt1m commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5906,4 +5943,81 @@ void RenderBox::removeShapeOutsideInfo()
+// https://drafts.csswg.org/css-anchor-position-1/#anchor-center
+void RenderBox::computeAnchorCenteredPosition(LogicalExtentComputedValues& computedValues, RenderBoxModelObject* defaultAnchorBox, Length logicalLeftLength, Length logicalRightLength, LayoutUnit containerLogicalWidth, bool computeHorizontally) const
+{
+    const RenderBlock* closestCommonContainer = findClosestCommonContainer(*this, *defaultAnchorBox);
```
```suggestion
    CheckedPtr closestCommonContainer = findClosestCommonContainer(*this, *defaultAnchorBox);
```

nt1m commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -4418,6 +4435,19 @@ void RenderBox::computePositionedLogicalHeight(LogicalExtentComputedValues& comp
+    bool alignsToAnchorCenter = container()->isHorizontalWritingMode() != isHorizontalWritingMode()
+        ? style().justifySelf().position() == ItemPosition::AnchorCenter
+        : style().alignSelf().position() == ItemPosition::AnchorCenter;
+    RenderBoxModelObject* defaultAnchorBox = alignsToAnchorCenter ? dynamicDowncast<RenderBoxModelObject>(defaultAnchorRenderer()) : nullptr;
```
```suggestion
    CheckedPtr defaultAnchorBox = alignsToAnchorCenter ? dynamicDowncast<RenderBoxModelObject>(defaultAnchorRenderer()) : nullptr;
```

Scony replied:
`defaultAnchorBox` can be `nullptr` in the above, so `CheckedPtr` will trigger assertion, won't it?

nt1m replied:
I don't think that's how CheckedPtr works, but it should be `auto*` at the very least.

alanbaradlay replied:
CheckedPtr as the name implies can hold a nullptr value (look at its c'tors) Maybe it's CheckedRef you were thinking of?

alanbaradlay replied:
unconditionally derefing it may trigger assertion.

Scony replied:
Right, I confused it with `CheckedRef` - sorry for confusion.

nt1m commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -4005,6 +4006,19 @@ void RenderBox::computePositionedLogicalWidth(LogicalExtentComputedValues& compu
+    bool alignsToAnchorCenter = container()->isHorizontalWritingMode() != isHorizontalWritingMode()
+        ? style().alignSelf().position() == ItemPosition::AnchorCenter
+        : style().justifySelf().position() == ItemPosition::AnchorCenter;
+    RenderBoxModelObject* defaultAnchorBox = alignsToAnchorCenter ? dynamicDowncast<RenderBoxModelObject>(defaultAnchorRenderer()) : nullptr;
```
```suggestion
    CheckedPtr defaultAnchorBox = alignsToAnchorCenter ? dynamicDowncast<RenderBoxModelObject>(defaultAnchorRenderer()) : nullptr;
```
same here fwiw

anttijk commented on Source/WebCore/rendering/RenderElement.h:
```diff
@@ -350,6 +350,9 @@ class RenderElement : public RenderObject {
     inline bool shouldApplySizeOrStyleContainment(bool) const;
 
+    Element* defaultAnchor() const;
+    RenderElement* defaultAnchorRenderer() const;
```
Can these return const Element*/const RenderElement* and avoid implicitly breaking const?

anttijk commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -4035,6 +4036,19 @@ void RenderBox::computePositionedLogicalWidth(LogicalExtentComputedValues& compu
+    const bool alignsToAnchorCenter = container()->isHorizontalWritingMode() != isHorizontalWritingMode()
+        ? style().alignSelf().position() == ItemPosition::AnchorCenter
+        : style().justifySelf().position() == ItemPosition::AnchorCenter;
+    CheckedPtr defaultAnchorBox = alignsToAnchorCenter ? dynamicDowncast<RenderBoxModelObject>(defaultAnchorRenderer()) : nullptr;
```
I'd write this as a lambda 
```
auto boxForAnchorCenter = [&]() -> CheckedPtr<RenderBoxModelObject> { 
    ...
};

anttijk commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -4035,6 +4036,19 @@ void RenderBox::computePositionedLogicalWidth(LogicalExtentComputedValues& compu
+        : style().justifySelf().position() == ItemPosition::AnchorCenter;
+    CheckedPtr defaultAnchorBox = alignsToAnchorCenter ? dynamicDowncast<RenderBoxModelObject>(defaultAnchorRenderer()) : nullptr;
+    // Any auto inset properties resolve to 0 if the box is absolutely positioned and does have a default anchor box.
+    if (alignsToAnchorCenter && defaultAnchorBox) {
```
It is sufficient to test `defaultAnchorBox` (here and below)

anttijk commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -4458,6 +4475,19 @@ void RenderBox::computePositionedLogicalHeight(LogicalExtentComputedValues& comp
+    const bool alignsToAnchorCenter = container()->isHorizontalWritingMode() != isHorizontalWritingMode()
+        ? style().justifySelf().position() == ItemPosition::AnchorCenter
+        : style().alignSelf().position() == ItemPosition::AnchorCenter;
+    CheckedPtr defaultAnchorBox = alignsToAnchorCenter ? dynamicDowncast<RenderBoxModelObject>(defaultAnchorRenderer()) : nullptr;
```
same here

Scony committed:
[css-anchor-position-1] Implement `anchor-center` alignment value
https://bugs.webkit.org/show_bug.cgi?id=275451

Reviewed by Antti Koivisto.

This change implements `anchor-center` CSS value as specified in:
https://drafts.csswg.org/css-anchor-position-1/#anchor-center

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-htb-htb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-htb-vrl-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-vrl-htb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-vrl-vrl-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computePositionedLogicalWidth const):
(WebCore::RenderBox::computePositionedLogicalHeight const):
(WebCore::isObjectAncestorContainerOf):
(WebCore::findClosestCommonContainer):
(WebCore::RenderBox::computeAnchorCenteredPosition const):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::defaultAnchor const):
(WebCore::RenderElement::defaultAnchorRenderer const):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::alignmentOffset):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::columnAxisPositionForGridItem const):
(WebCore::RenderGrid::rowAxisPositionForGridItem const):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::computeAnchorRectRelativeToContainingBlock):
(WebCore::Style::computeInsetValue):
(WebCore::Style::computeAnchorRectRelativeToContainingBlock): Deleted.
* Source/WebCore/style/AnchorPositionEvaluator.h:

Canonical link: https://commits.webkit.org/287144@main
